### ISSUE TITLE
`obj_size` returns different values before and after accessing an element

### ISSUE THREAD

**Issue by bersbersbers** [r-lib/lobstr#59]
I am chasing memory leaks and/or high memory fragmentation in a Shiny app. I have noticed that after closing the Shiny app and returning to the R prompt, `gc()` still reports some 3.8 GB of memory used even after emptying all environments (except `baseenv()`) and unloading all namespaces, which is quite a bit. (For context, Linux `top` reports some 11 GB of virtual memory used.)

I have started looking at the remaining elements in `baseenv()`, and found that almost all of the memory is in `baseenv()$.__S3MethodsTable__.` (disclaimer: I have *no idea* what this is for and why this could ever be this big):

```r
lobstr::obj_size(baseenv()$.__S3MethodsTable__.)
# 3,314,445,280 B
```

So I went ahead to see which of its elements is the culprit:

```r
sizes <- lapply(names(baseenv()$.__S3MethodsTable__.), \(name) lobstr::obj_size(baseenv()$.__S3MethodsTable__.[[name]]))
names(sizes) <- names(baseenv()$.__S3MethodsTable__.)
sizes <- sizes[order(unlist(sizes))]
```

To my astonishment, after I have done this, the reported size of `baseenv()$.__S3MethodsTable__.` has drastically decreased:

```r
lobstr::obj_size(baseenv()$.__S3MethodsTable__.)
# 157,085,600 B
```

The same happens when I inspect members of `baseenv()$.__S3MethodsTable__.`:

```r
sizes
# $`[[.SQL`
# 93,406,504 B
# 
# $`[.SQL`
# 93,406,504 B
# 
# $toString.Id
# 93,411,712 B

lobstr::obj_size(baseenv()$.__S3MethodsTable__.$toString.Id)
# 93,411,712 B
baseenv()$.__S3MethodsTable__.$toString.Id
# function(x, ...) {
#   paste0("<Id> ", paste0(names(x@name), " = ", x@name, collapse = ", "))
# }
# <bytecode: 0x55a1f4ea97c8>
# <environment: namespace:DBI>
lobstr::obj_size(baseenv()$.__S3MethodsTable__.$toString.Id)
# 297,264 B
```

What is going on here? I almost feel like the memory leak is playing whac-a-mole. Is this a bug in `obj_size`, expected behavior, or actually a sign of memory leaks?

<s>Interestingly, in a clean R session, this element has yet another size:</s> (That was comparing Windows with Linux - it's consistent on Linux.)

<s>

```r
library(DBI)
baseenv()$.__S3MethodsTable__.$toString.Id
# function (x, ...)
# {
#     paste0("<Id> ", paste0(names(x@name), " = ", x@name, collapse = ", "))
# }
# <bytecode: 0x00000000194d22f0>
# <environment: namespace:DBI>
lobstr::obj_size(baseenv()$.__S3MethodsTable__.$toString.Id)
# 4,320 B
```

</s>

**Comment by hadley** [comment]
Hmmm, interesting. I see the opposite:

``` r
lobstr::obj_size(baseenv()$.__S3MethodsTable__.)
#> 6,307,928 B
x <- sort(lobstr::obj_sizes(!!!as.list(baseenv()$.__S3MethodsTable__.)))
lobstr::obj_size(baseenv()$.__S3MethodsTable__.)
#> 12,894,608 B
```

<sup>Created on 2022-06-17 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

I think this environment is the registry of S3 methods for all base functions, but it is surprising that it is so large for you. 
I _think_ it's only affected by the packages you load, so you might try creating a reprex by extracting just the `library()` statements from your shiny app.

The size of the individual methods is also surprising, but I can't reproduce it:

``` r
library(DBI)
library(tidyverse)

lobstr::obj_size(baseenv()$.__S3MethodsTable__.$toString.Id)
#> 4,320 B
baseenv()$.__S3MethodsTable__.$toString.Id
#> function (x, ...) 
#> {
#>     paste0("<Id> ", paste0(names(x@name), " = ", x@name, collapse = ", "))
#> }
#> <bytecode: 0x7fbcb887ca90>
#> <environment: namespace:DBI>
lobstr::obj_size(baseenv()$.__S3MethodsTable__.$toString.Id)
#> 4,320 B
```

<sup>Created on 2022-06-17 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

If possible, I'd also recommend you trying this code in the latest version of R, in case this is a some buglet in base R.

**Comment by bersbersbers** [comment]
This issue is a few months old, so I don't have anything to reproduce my original results, unfortunately.

What I have found out in the meantime, however, is that the "memory leak" I was after turned out to be a couple of paired `saveRDS` and `readRDS` (of the same file, respectively), and these `saveRDS` operations involved `ggplot2` objects, and hence the complete environments. I cannot say for sure whether that actually consumed a lot of memory or just made it appear so, but performance was certainly not good. When I replaced `readRDS` operations by reading the plots from an environment that I pass around, most memory and runtime issues disappeared. (Any they completely disappeared when I also removed the `saveRDS` calls, after reading your own https://github.com/tidyverse/ggplot2/issues/3994#issuecomment-1142116587).

One (very much uninformed) hypothesis that I have is that apparent memory consumption goes down when inspection of the members of some `__S3MethodsTable__` reveals that many duplicates of each other that use the same memory, but again, that is a very wild guess.
